actionpack/lib/action_controller/metal/strong_parameters.rb

```diff
@@ -292,11 +292,19 @@ def to_hash
       if self.class.raise_on_unfiltered_parameters
         to_h.to_hash
       else
-        message = <<-DEPRECATE.squish
-          #to_hash unexpectedly ignores parameter filtering, and will change to enforce it in Rails 5.1.
-          Enable `raise_on_unfiltered_parameters` to respect parameter filtering, which is the default
-          in new applications. For the existing deprecated behaviour, call #to_unsafe_h instead.
-        DEPRECATE
+        if permitted?
+          message = <<-DEPRECATE.squish
```

Comment:
Those two messages are saying the same thing to me. I agree we need to differentiate both cases for pratical reasons. Maybe

```
#to_hash was called in filtered parameters. This method unexpectedly ignores parameter filtering, and will change to enforce it in Rails 5.1.
Enable `raise_on_unfiltered_parameters` to respect parameter filtering, which is the default
in new applications, and prevent regressions on the filtering. For the existing deprecated behaviour, call #to_unsafe_h instead.
```

in permitted instances and:

```
#to_hash was called in filtered parameters. This method unexpectedly ignores parameter filtering, and will change to enforce it in Rails 5.1.
Filter it before calling `#to_hash` or call #to_unsafe_h instead, for the existing deprecated behaviour.
```

Reply:
"prevent regressions on the filtering"? :confused:

"Filter it before calling `#to_hash`" from context, I assume that means "call `permit`" -- but that won't clear the deprecation: it'll just move them to the other warning.

How about:

> #to_hash unexpectedly ignores parameter filtering, and calling it on unfiltered parameters will raise an exception in Rails 5.1. Filter to the expected parameters with #permit, then call #to_h for the new behavior. For the existing deprecated behaviour, call #to_unsafe_h instead.

and

> #to_hash unexpectedly ignores parameter filtering, and will change to enforce it in Rails 5.1. Enable `raise_on_unfiltered_parameters` or call #to_h for the new behavior. For the existing deprecated behaviour, call #to_unsafe_h instead.

Reply:
Those give you your different wordings for the two situations, but the "what you should do now" part sure sounds the same to me. 

Does the motivation for this change boil down to a central misunderstanding that "the case where parameters are already permitted [..] isn't changing"? Maybe that's the part we need to emphasise more strongly?

Reply:
The motivation comes from not being able to distuiguich between:

```ruby
params = Parameter.new(foo: "foo", bar: "bar")
params.permit(:foo).to_hash # => { foo: "foo" }
```

and

```ruby
params = Parameter.new(foo: "foo", bar: "bar")
params.to_hash # => { foo: "foo", bar: "bar"}
```

The first one if fine to enable that config and keep the code as it is. In the second one it will cause an exception, so we need to change the code.

Reply:
> #to_hash unexpectedly ignores parameter filtering, and calling it on unfiltered parameters will raise an exception in Rails 5.1. Filter to the expected parameters with #permit, then call #to_h for the new behavior. For the existing deprecated behaviour, call #to_unsafe_h instead.

The deprecation for the unfiltered parameters message you suggested make it very clear on the consequences on leaving this as is if someone wants to upgrade rails as well as what action to take to solve it. And as an effect the other deprecation message make it easier to distinguished

I'm not sure if emphasizing the "to_hash on permitted params is fine, enable the flag" would help more.

Reply:
> ```ruby
> params = Parameter.new(foo: "foo", bar: "bar")
> params.permit(:foo).to_hash # => { foo: "foo" }
> ```

Okay, I'm confused. I thought the current behaviour was:

```ruby
params = Parameter.new(foo: "foo", bar: "bar")
params.permit(:foo).to_hash # => { foo: "foo", bar: "bar" }
```

.. and the point of the deprecation message is that it's *not* fine, because it's going to *change* to your example (only `foo`).

If *all* the parameters present have been permitted, then it's fine -- and we could (should?) skip the warning entirely. But if only some are, `to_hash` is going to give a different result when the config is switched, and that's what we're warning about: it's worse than something that'll start raising: it'll keep working, while silently behaving differently.

Reply:
When you call `permit` a new instance of `Parameter` is returned only with the permitted parameters so `to_hash` on it will only return the permitted parameters.

https://github.com/rails/rails/blob/5-0-stable/actionpack/test/controller/parameters/parameters_permit_test.rb#L79-L84

In this case it is fine to call `to_hash` in that object.

I'm fine to skip the warning entirely, but I'd prefer to encourage people to enable that config.

Reply:
The behaviour of `Parameter.new(foo: "foo", bar: "bar").permit(:foo).to_hash` isn't changing, it already returns `{"foo"=>"foo"}` in rails 5.0, which is why I said earlier

> A deprecation warning in the case where parameters are already permitted is kind of useless to us

Given that, could we just warn on unpermitted parameters?  Since the behaviour for permitted parameters isn't changing.

Reply:
Oh! I thought `permit` calls were additive, and the whole issue we were fixing was that `to_hash` ignored filtering that had been specified.

Sorry, I obviously misunderstood "ignores parameter filtering", possibly since we first introduced it. 😵

In that case, I don't understand why the warning is there, and agree we should remove it. Deprecations should be actionable, at the call site that's triggering them. 

"Hey, if you were making this call on a *different* object, it might change in a future version" isn't an appropriate place to advertise a config setting that would have global effect, much as I'd like everyone to enable it. The whole point of the deprecation [on unpermitted params, where we should keep it] is to warn people that changing the config will affect the behaviour of their application: until they've audited all such call sites, they should *not* enable it.

Reply:
:+1:. Let's remove them.